distsql: TestClusterFlow span use-after-finish

```
I161014 14:51:50.527167 11 storage/engine/rocksdb.go:396  closing in-memory rocksdb instance
panic: span used after call to Finish():
&{tracer:<nil> event:0x7f5a00 Mutex:{state:0 sema:0} raw:{Context:{TraceID:6227312592976378248 SpanID:7151576361212056128 Sampled:false Baggage:map[]} ParentSpanID:8033824228644420073 Operation:flow Start:{sec:63612053510 nsec:493973503 loc:0x23e27c0} Duration:29156022 Tags:map[] Logs:[]} numDroppedLogs:0}

goroutine 289 [running]:
panic(0x153dda0, 0xc420ab1aa0)
    /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/opentracing/basictracer-go.(*spanImpl).maybeAssertSanityLocked(0xc4206348c0)
    /go/src/github.com/opentracing/basictracer-go/debug.go:31 +0x341
github.com/opentracing/basictracer-go.(*spanImpl).Lock(0xc4206348c0)
    /go/src/github.com/opentracing/basictracer-go/debug.go:25 +0x3f
github.com/opentracing/basictracer-go.(*spanImpl).Log(0xc4206348c0, 0x0, 0x0, 0x0, 0xc420ac2dc0, 0x3c, 0x0, 0x0)
    /go/src/github.com/opentracing/basictracer-go/span.go:149 +0x9a
github.com/opentracing/basictracer-go.(*spanImpl).LogEvent(0xc4206348c0, 0xc420ac2dc0, 0x3c)
    /go/src/github.com/opentracing/basictracer-go/span.go:137 +0x9e
github.com/cockroachdb/cockroach/pkg/util/log.eventInternal(0x7f3ee75e79c8, 0xc4201a3320, 0xc420ad0000, 0x16a5f54, 0x8, 0xc420509920, 0x3, 0x3)
    /go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:123 +0x3a4
github.com/cockroachdb/cockroach/pkg/util/log.addStructured(0x7f3ee75e79c8, 0xc4201a3320, 0xc400000001, 0x2, 0x16b1c5e, 0x13, 0x0, 0x0, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/util/log/structured.go:142 +0x25a
github.com/cockroachdb/cockroach/pkg/util/log.logDepth(0x7f3ee75e79c8, 0xc4201a3320, 0x1, 0xc400000001, 0x16b1c5e, 0x13, 0x0, 0x0, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:88 +0x8c
github.com/cockroachdb/cockroach/pkg/util/log.Infof(0x7f3ee75e79c8, 0xc4201a3320, 0x16b1c5e, 0x13, 0x0, 0x0, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:96 +0x7e
github.com/cockroachdb/cockroach/pkg/sql/distsql.ProcessInboundStream.func1(0x0, 0x0, 0x0, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/inbound.go:42 +0x1ef
github.com/cockroachdb/cockroach/pkg/sql/distsql.ProcessInboundStream(0xc42001a630, 0x209eaa0, 0xc420b10190, 0x0, 0x208d3a0, 0xc420180d40, 0x10, 0x10)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/inbound.go:89 +0x4c2
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).flowStreamInt(0xc420658180, 0x209eaa0, 0xc420b10190, 0xc420940c68, 0x5d2268)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:185 +0x1f6
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).FlowStream(0xc420658180, 0x209eaa0, 0xc420b10190, 0xc420940ca8, 0x20)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:190 +0x43
github.com/cockroachdb/cockroach/pkg/sql/distsql._DistSQL_FlowStream_Handler(0x1606b40, 0xc420658180, 0x209b620, 0xc42096e120, 0x8000, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/api.pb.go:264 +0xbb
google.golang.org/grpc.(*Server).processStreamingRPC(0xc42041d180, 0x209eec0, 0xc420526750, 0xc420510690, 0xc42040ef60, 0x2075be0, 0x0, 0x0, 0x0)
    /go/src/google.golang.org/grpc/server.go:689 +0x702
google.golang.org/grpc.(*Server).handleStream(0xc42041d180, 0x209eec0, 0xc420526750, 0xc420510690, 0x0)
    /go/src/google.golang.org/grpc/server.go:773 +0xc36
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc420aeffe0, 0xc42041d180, 0x209eec0, 0xc420526750, 0xc420510690)
    /go/src/google.golang.org/grpc/server.go:422 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
    /go/src/google.golang.org/grpc/server.go:423 +0xa3

```

https://teamcity.cockroachdb.com/viewLog.html?buildId=31876&buildTypeId=Cockroach_UnitTests

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -602,6 +602,8 @@ func (s *Server) Start(ctx context.Context) error {
 	}
 	sql.NewSchemaChangeManager(testingKnobs, *s.db, s.gossip, s.leaseMgr).Start(s.stopper)
 
+	s.distSQLServer.Start()
+
 	log.Infof(ctx, "starting %s server at %s", s.cfg.HTTPRequestScheme(), unresolvedHTTPAddr)
 	log.Infof(ctx, "starting grpc/postgres server at %s", unresolvedListenAddr)
 	log.Infof(ctx, "advertising CockroachDB node at %s", unresolvedAdvertAddr)

--- a/pkg/sql/dist_sql_node.go
+++ b/pkg/sql/dist_sql_node.go
@@ -120,7 +120,7 @@ func (n *distSQLNode) Next() (bool, error) {
 		if n.syncMode {
 			n.flow.RunSync()
 		} else {
-			n.flow.Start()
+			n.flow.Start(func() {})
 		}
 		n.flowStarted = true
 	}

--- a/pkg/sql/distsql/flow.go
+++ b/pkg/sql/distsql/flow.go
@@ -64,16 +64,27 @@ const (
 type Flow struct {
 	FlowCtx
 
-	flowRegistry       *flowRegistry
+	flowRegistry *flowRegistry
+	processors   []processor
+	outboxes     []*outbox
+	// simpleFlowConsumer is a special outbox which instead of sending rows to
+	// another host, returns them directly (as a result to a SetupSimpleFlow RPC,
+	// or to the local host).
 	simpleFlowConsumer RowReceiver
-	waitGroup          sync.WaitGroup
-	processors         []processor
-	outboxes           []*outbox
 
-	// inboundStreams are streams that receive data from other hosts, through
-	// the FlowStream API.
-	inboundStreams map[StreamID]RowReceiver
-	localStreams   map[StreamID]RowReceiver
+	localStreams map[StreamID]RowReceiver
+
+	// inboundStreams are streams that receive data from other hosts; this map
+	// is to be passed to flowRegistry.RegisterFlow.
+	inboundStreams map[StreamID]*inboundStreamInfo
+
+	// waitGroup is used to wait for async components of the flow:
+	//  - processors
+	//  - inbound streams
+	//  - outboxes
+	waitGroup sync.WaitGroup
+
+	doneFn func()
 
 	status flowStatus
 }
@@ -83,12 +94,13 @@ func newFlow(flowCtx FlowCtx, flowReg *flowRegistry, simpleFlowConsumer RowRecei
 		panic("flow context has no span")
 	}
 	flowCtx.Context = log.WithLogTagStr(flowCtx.Context, "f", flowCtx.id.Short())
-	return &Flow{
+	f := &Flow{
 		FlowCtx:            flowCtx,
 		flowRegistry:       flowReg,
 		simpleFlowConsumer: simpleFlowConsumer,
-		status:             FlowNotStarted,
 	}
+	f.status = FlowNotStarted
+	return f
 }
 
 // setupInboundStream adds a stream to the stream map (inboundStreams or
@@ -103,12 +115,12 @@ func (f *Flow) setupInboundStream(spec StreamEndpointSpec, receiver RowReceiver)
 			return errors.Errorf("inbound stream %d has multiple consumers", sid)
 		}
 		if f.inboundStreams == nil {
-			f.inboundStreams = make(map[StreamID]RowReceiver)
+			f.inboundStreams = make(map[StreamID]*inboundStreamInfo)
 		}
 		if log.V(2) {
 			log.Infof(f.FlowCtx.Context, "set up inbound stream %d", sid)
 		}
-		f.inboundStreams[sid] = receiver
+		f.inboundStreams[sid] = &inboundStreamInfo{receiver: receiver, waitGroup: &f.waitGroup}
 		return nil
 	}
 	if _, found := f.localStreams[sid]; found {
@@ -295,26 +307,20 @@ func (f *Flow) setupFlow(spec *FlowSpec) error {
 	return nil
 }
 
-func (f *Flow) getInboundStream(sid StreamID) (RowReceiver, error) {
-	// TODO(radu): detect if we connect to a stream multiple times.
-	recv, ok := f.inboundStreams[sid]
-	if !ok {
-		return nil, errors.Errorf("no inbound stream %d for flow %s", sid, f.id)
-	}
-	return recv, nil
-}
-
 // Start starts the flow (each processor runs in their own goroutine).
-func (f *Flow) Start() {
+func (f *Flow) Start(doneFn func()) {
+	f.doneFn = doneFn
 	log.VEventf(
 		f.Context, 1, "starting (%d processors, %d outboxes)", len(f.outboxes), len(f.processors),
 	)
 	f.status = FlowRunning
-	f.flowRegistry.RegisterFlow(f.id, f)
+	f.flowRegistry.RegisterFlow(f.id, f, f.inboundStreams)
+	f.waitGroup.Add(len(f.inboundStreams))
+	f.waitGroup.Add(len(f.outboxes))
+	f.waitGroup.Add(len(f.processors))
 	for _, o := range f.outboxes {
 		o.start(&f.waitGroup)
 	}
-	f.waitGroup.Add(len(f.processors))
 	for _, p := range f.processors {
 		go p.Run(&f.waitGroup)
 	}
@@ -340,6 +346,8 @@ func (f *Flow) Cleanup() {
 		f.flowRegistry.UnregisterFlow(f.id)
 	}
 	f.status = FlowFinished
+	f.doneFn()
+	f.doneFn = nil
 }
 
 // RunSync runs the processors in the flow in order (serially), in the same

--- a/pkg/sql/distsql/flow_registry.go
+++ b/pkg/sql/distsql/flow_registry.go
@@ -18,10 +18,39 @@ package distsql
 
 import (
 	"fmt"
+	"sync"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/pkg/errors"
 )
+
+// flowStreamTimeout is the amount of time incoming streams wait for a flow to
+// be set up before erroring out.
+const flowStreamTimeout time.Duration = 4000 * time.Millisecond
+
+// inboundStreamInfo represents the endpoint where a data stream from another
+// node connects to a flow. The external node initiates this process through a
+// FlowStream RPC, which uses (*Flow).connectInboundStream() to associate the
+// stream to a receiver to push rows to.
+type inboundStreamInfo struct {
+	// receiver is the entity that will receive rows from another host, which is
+	// part of a processor (normally an input synchronizer).
+	//
+	// During a FlowStream RPC, rows are pushed to this entity using the
+	// RowReceiver interface.
+	receiver  RowReceiver
+	connected bool
+	// if set, indicates that we waited too long for an inbound connection.
+	timedOut bool
+	// finished is set if we have signaled that the stream is done transferring
+	// rows (to the flow's wait group).
+	finished bool
+
+	// waitGroup to signal on when finished.
+	waitGroup *sync.WaitGroup
+}
 
 // flowEntry is a structure associated with a (potential) flow.
 // All fields are protected by the flowRegistry mutex.
@@ -36,12 +65,20 @@ type flowEntry struct {
 	refCount int
 
 	flow *Flow
+
+	// inboundStreams are streams that receive data from other hosts, through
+	// the FlowStream API.
+	inboundStreams map[StreamID]*inboundStreamInfo
+
+	// streamTimer is a timer that fires after a timeout and verifies that all
+	// inbound streams have been connected.
+	streamTimer *time.Timer
 }
 
 // flowRegistry allows clients to look up flows by ID and to wait for flows to
 // be registered. Multiple clients can wait concurrently for the same flow.
 type flowRegistry struct {
-	mu    syncutil.Mutex
+	syncutil.Mutex
 	flows map[FlowID]*flowEntry
 }
 
@@ -83,9 +120,11 @@ func (fr *flowRegistry) releaseEntryLocked(id FlowID) {
 // calls that are waiting for this flow are woken up.
 // It is expected that UnregisterFlow will be called at some point to remove the
 // flow from the registry.
-func (fr *flowRegistry) RegisterFlow(id FlowID, f *Flow) {
-	fr.mu.Lock()
-	defer fr.mu.Unlock()
+func (fr *flowRegistry) RegisterFlow(
+	id FlowID, f *Flow, inboundStreams map[StreamID]*inboundStreamInfo,
+) {
+	fr.Lock()
+	defer fr.Unlock()
 	entry := fr.getEntryLocked(id)
 	if entry.flow != nil {
 		panic("flow already registered")
@@ -93,29 +132,59 @@ func (fr *flowRegistry) RegisterFlow(id FlowID, f *Flow) {
 	// Take a reference that will be removed by UnregisterFlow.
 	entry.refCount++
 	entry.flow = f
+	entry.inboundStreams = inboundStreams
 	// If there are any waiters, wake them up by closing waitCh.
 	if entry.waitCh != nil {
 		close(entry.waitCh)
+	}
+
+	if len(inboundStreams) > 0 {
+		// Set up a function to time out inbound streams after a while.
+		entry.streamTimer = time.AfterFunc(flowStreamTimeout, func() {
+			fr.Lock()
+			defer fr.Unlock()
+			numTimedOut := 0
+			for _, is := range entry.inboundStreams {
+				if is.timedOut {
+					panic("stream already timed out")
+				}
+				if !is.connected {
+					is.timedOut = true
+					numTimedOut++
+					is.receiver.Close(errors.Errorf("no inbound stream connection"))
+					fr.finishInboundStreamLocked(is)
+				}
+			}
+			if numTimedOut != 0 {
+				log.Errorf(entry.flow.Context, "%d inbound streams timed out after %s; stopping flow",
+					numTimedOut, flowStreamTimeout)
+			}
+		})
 	}
 }
 
 // UnregisterFlow removes a flow from the registry. Any subsequent LookupFlow
 // calls will time out.
 func (fr *flowRegistry) UnregisterFlow(id FlowID) {
-	fr.mu.Lock()
+	fr.Lock()
+	entry := fr.flows[id]
+	if entry.streamTimer != nil {
+		entry.streamTimer.Stop()
+		entry.streamTimer = nil
+	}
 	fr.releaseEntryLocked(id)
-	fr.mu.Unlock()
+	fr.Unlock()
 }
 
-// LookupFlow returns the registered flow with the given ID. If no such flow is
-// registered, waits until it gets registered - up to the given timeout. If the
-// timeout elapses, returns nil.
-func (fr *flowRegistry) LookupFlow(id FlowID, timeout time.Duration) *Flow {
-	fr.mu.Lock()
-	defer fr.mu.Unlock()
+// waitForFlowLocked returns the flowEntry of a registered flow with the given
+// ID. If no such flow is registered, waits until it gets registered - up to the
+// given timeout. If the timeout elapses, returns nil. It should only be called
+// while holding the mutex. The mutex is temporarily unlocked if we need to
+// wait.
+func (fr *flowRegistry) waitForFlowLocked(id FlowID, timeout time.Duration) *flowEntry {
 	entry := fr.getEntryLocked(id)
 	if entry.flow != nil {
-		return entry.flow
+		return entry
 	}
 
 	// Flow not registered (at least not yet).
@@ -127,7 +196,7 @@ func (fr *flowRegistry) LookupFlow(id FlowID, timeout time.Duration) *Flow {
 		entry.waitCh = make(chan struct{})
 	}
 	entry.refCount++
-	fr.mu.Unlock()
+	fr.Unlock()
 
 	// Wait until waitCh gets closed or the timeout elapses.
 	select {
@@ -135,9 +204,71 @@ func (fr *flowRegistry) LookupFlow(id FlowID, timeout time.Duration) *Flow {
 	case <-time.After(timeout):
 	}
 
-	fr.mu.Lock()
+	fr.Lock()
 
 	fr.releaseEntryLocked(id)
+	if entry.flow == nil {
+		return nil
+	}
 
+	return entry
+}
+
+// LookupFlow returns the registered flow with the given ID. If no such flow is
+// registered, waits until it gets registered - up to the given timeout. If the
+// timeout elapses, returns nil.
+func (fr *flowRegistry) LookupFlow(id FlowID, timeout time.Duration) *Flow {
+	fr.Lock()
+	defer fr.Unlock()
+	entry := fr.waitForFlowLocked(id, timeout)
+	if entry == nil {
+		return nil
+	}
 	return entry.flow
+}
+
+// ConnectInboundStream finds the inboundStreamInfo for the given ID and marks it
+// as connected. FinishInboundStream must be called after the rows are
+// transferred.
+func (fr *flowRegistry) ConnectInboundStream(
+	flowID FlowID, streamID StreamID,
+) (*Flow, *inboundStreamInfo, error) {
+	fr.Lock()
+	defer fr.Unlock()
+	entry := fr.waitForFlowLocked(flowID, flowStreamTimeout)
+	if entry == nil {
+		return nil, nil, errors.Errorf("flow %s not found", flowID)
+	}
+
+	s, ok := entry.inboundStreams[streamID]
+	if !ok {
+		return nil, nil, errors.Errorf("flow %s: no inbound stream %d", flowID, streamID)
+	}
+	if s.connected {
+		return nil, nil, errors.Errorf("flow %s: inbound stream %d already connected", flowID, streamID)
+	}
+	if s.timedOut {
+		return nil, nil, errors.Errorf("flow %s: inbound stream %d came too late", flowID, streamID)
+	}
+	s.connected = true
+	return entry.flow, s, nil
+}
+
+func (fr *flowRegistry) finishInboundStreamLocked(is *inboundStreamInfo) {
+	if !is.connected && !is.timedOut {
+		panic("finishing inbound stream that didn't connect or time out")
+	}
+	if is.finished {
+		panic("double finish")
+	}
+	is.finished = true
+	is.waitGroup.Done()
+}
+
+// FinishInboundStream is to be called when we are done transferring rows for a
+// stream previously connected via ConnectInboundStream.
+func (fr *flowRegistry) FinishInboundStream(is *inboundStreamInfo) {
+	fr.Lock()
+	defer fr.Unlock()
+	fr.finishInboundStreamLocked(is)
 }

--- a/pkg/sql/distsql/flow_registry_test.go
+++ b/pkg/sql/distsql/flow_registry_test.go
@@ -48,7 +48,7 @@ func TestFlowRegistry(t *testing.T) {
 		t.Error("looked up unregistered flow")
 	}
 
-	reg.RegisterFlow(id1, f1)
+	reg.RegisterFlow(id1, f1, nil)
 
 	if f := reg.LookupFlow(id1, 0); f != f1 {
 		t.Error("couldn't lookup previously registered flow")
@@ -64,7 +64,7 @@ func TestFlowRegistry(t *testing.T) {
 
 	go func() {
 		time.Sleep(jiffy)
-		reg.RegisterFlow(id1, f1)
+		reg.RegisterFlow(id1, f1, nil)
 	}()
 
 	if f := reg.LookupFlow(id1, 10*jiffy); f != f1 {
@@ -95,7 +95,7 @@ func TestFlowRegistry(t *testing.T) {
 	}()
 
 	time.Sleep(jiffy)
-	reg.RegisterFlow(id2, f2)
+	reg.RegisterFlow(id2, f2, nil)
 	wg.Wait()
 
 	// -- Multiple lookups, with the first one failing. --
@@ -120,6 +120,6 @@ func TestFlowRegistry(t *testing.T) {
 	}()
 
 	wg1.Wait()
-	reg.RegisterFlow(id3, f3)
+	reg.RegisterFlow(id3, f3, nil)
 	wg2.Wait()
 }

--- a/pkg/sql/distsql/flow_scheduler.go
+++ b/pkg/sql/distsql/flow_scheduler.go
@@ -1,0 +1,119 @@
+// Copyright 2016 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Radu Berinde (radu@cockroachlabs.com)
+
+package distsql
+
+import (
+	"container/list"
+
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+)
+
+const maxRunningFlows = 500
+const flowDoneChanSize = 8
+
+// flowScheduler manages running flows and decides when to queue and when to
+// start flows. The main interface it presents is ScheduleFlows, which passes a
+// flow to be run.
+type flowScheduler struct {
+	log.AmbientContext
+	stopper    *stop.Stopper
+	flowDoneCh chan *Flow
+
+	mu struct {
+		syncutil.Mutex
+		numRunning int
+		queue      *list.List
+	}
+}
+
+func newFlowScheduler(ambient log.AmbientContext, stopper *stop.Stopper) *flowScheduler {
+	fs := &flowScheduler{
+		AmbientContext: ambient,
+		stopper:        stopper,
+		flowDoneCh:     make(chan *Flow, flowDoneChanSize),
+	}
+	fs.mu.queue = list.New()
+	return fs
+}
+
+func (fs *flowScheduler) canRunFlow(_ *Flow) bool {
+	// TODO(radu): we will have more complex resource accounting (like memory).
+	// For now we just limit the number of concurrent flows.
+	return fs.mu.numRunning < maxRunningFlows
+}
+
+// runFlowNow starts the given flow; does not wait for the flow to complete.
+func (fs *flowScheduler) runFlowNow(f *Flow) {
+	fs.mu.numRunning++
+	f.Start(func() { fs.flowDoneCh <- f })
+	// TODO(radu): we could replace the WaitGroup with a structure that keeps a
+	// refcount and automatically runs Cleanup() when the count reaches 0.
+	go func() {
+		f.Wait()
+		f.Cleanup()
+	}()
+}
+
+// ScheduleFlow is the main interface of the flow scheduler: it runs or enqueues
+// the given flow.
+func (fs *flowScheduler) ScheduleFlow(f *Flow) error {
+	return fs.stopper.RunTask(func() {
+		fs.mu.Lock()
+		defer fs.mu.Unlock()
+
+		if fs.canRunFlow(f) {
+			fs.runFlowNow(f)
+		} else {
+			fs.mu.queue.PushBack(f)
+		}
+	})
+}
+
+// Start launches the main loop of the scheduler.
+func (fs *flowScheduler) Start() {
+	fs.stopper.RunWorker(func() {
+		stopped := false
+		fs.mu.Lock()
+		defer fs.mu.Unlock()
+
+		for {
+			if stopped && fs.mu.numRunning == 0 {
+				// TODO(radu): somehow error out the flows that are still in the queue.
+				return
+			}
+			fs.mu.Unlock()
+			select {
+			case <-fs.flowDoneCh:
+				fs.mu.Lock()
+				fs.mu.numRunning--
+				if !stopped {
+					if frElem := fs.mu.queue.Front(); frElem != nil {
+						f := frElem.Value.(*Flow)
+						fs.mu.queue.Remove(frElem)
+						fs.runFlowNow(f)
+					}
+				}
+
+			case <-fs.stopper.ShouldStop():
+				fs.mu.Lock()
+				stopped = true
+			}
+		}
+	})
+}

--- a/pkg/sql/distsql/outbox.go
+++ b/pkg/sql/distsql/outbox.go
@@ -198,7 +198,6 @@ func (m *outbox) run() {
 }
 
 func (m *outbox) start(wg *sync.WaitGroup) {
-	wg.Add(1)
 	m.wg = wg
 	m.RowChannel.Init()
 	go m.run()

--- a/pkg/sql/distsql/server.go
+++ b/pkg/sql/distsql/server.go
@@ -18,7 +18,6 @@ package distsql
 
 import (
 	"io"
-	"time"
 
 	"golang.org/x/net/context"
 
@@ -46,13 +45,10 @@ type ServerConfig struct {
 // ServerImpl implements the server for the distributed SQL APIs.
 type ServerImpl struct {
 	ServerConfig
-	evalCtx      parser.EvalContext
-	flowRegistry *flowRegistry
+	evalCtx       parser.EvalContext
+	flowRegistry  *flowRegistry
+	flowScheduler *flowScheduler
 }
-
-// flowStreamTimeout is the amount of time incoming streams wait for a flow to
-// be set up before erroring out.
-const flowStreamTimeout time.Duration = 2000 * time.Millisecond
 
 var _ DistSQLServer = &ServerImpl{}
 
@@ -63,9 +59,15 @@ func NewServer(cfg ServerConfig) *ServerImpl {
 		evalCtx: parser.EvalContext{
 			ReCache: parser.NewRegexpCache(512),
 		},
-		flowRegistry: makeFlowRegistry(),
+		flowRegistry:  makeFlowRegistry(),
+		flowScheduler: newFlowScheduler(cfg.AmbientContext, cfg.Stopper),
 	}
 	return ds
+}
+
+// Start launches workers for the server.
+func (ds *ServerImpl) Start() {
+	ds.flowScheduler.Start()
 }
 
 func (ds *ServerImpl) setupTxn(ctx context.Context, txnProto *roachpb.Transaction) *client.Txn {
@@ -117,7 +119,7 @@ func (ds *ServerImpl) RunSimpleFlow(
 ) error {
 	// Set up the outgoing mailbox for the stream.
 	mbox := newOutboxSimpleFlowStream(stream)
-	ctx := ds.AnnotateCtx(context.TODO())
+	ctx := ds.AnnotateCtx(stream.Context())
 
 	f, err := ds.SetupSimpleFlow(ctx, req, mbox)
 	if err != nil {
@@ -127,8 +129,9 @@ func (ds *ServerImpl) RunSimpleFlow(
 	mbox.setFlowCtx(&f.FlowCtx)
 
 	if err := ds.Stopper.RunTask(func() {
+		f.waitGroup.Add(1)
 		mbox.start(&f.waitGroup)
-		f.Start()
+		f.Start(func() {})
 		f.Wait()
 		f.Cleanup()
 	}); err != nil {
@@ -146,13 +149,9 @@ func (ds *ServerImpl) SetupFlow(_ context.Context, req *SetupFlowRequest) (*Simp
 	if err != nil {
 		return nil, err
 	}
-	f.Start()
-	// TODO(radu): firing off a goroutine just to call Cleanup is temporary. We
-	// will have a flow scheduler that will be notified when the flow completes.
-	ds.Stopper.RunWorker(func() {
-		f.Wait()
-		f.Cleanup()
-	})
+	if err := ds.flowScheduler.ScheduleFlow(f); err != nil {
+		return nil, err
+	}
 	return &SimpleResponse{}, nil
 }
 
@@ -169,20 +168,17 @@ func (ds *ServerImpl) flowStreamInt(stream DistSQL_FlowStreamServer) error {
 		return errors.Errorf("no header in first message")
 	}
 	flowID := msg.Header.FlowID
-	f := ds.flowRegistry.LookupFlow(flowID, flowStreamTimeout)
-	if f == nil {
-		return errors.Errorf("flow %s not found", flowID)
-	}
-	rowChan, err := f.getInboundStream(msg.Header.StreamID)
+	f, streamInfo, err := ds.flowRegistry.ConnectInboundStream(flowID, msg.Header.StreamID)
 	if err != nil {
 		return err
 	}
-	return ProcessInboundStream(&f.FlowCtx, stream, msg, rowChan)
+	defer ds.flowRegistry.FinishInboundStream(streamInfo)
+	return ProcessInboundStream(&f.FlowCtx, stream, msg, streamInfo.receiver)
 }
 
 // FlowStream is part of the DistSQLServer interface.
 func (ds *ServerImpl) FlowStream(stream DistSQL_FlowStreamServer) error {
-	ctx := ds.AnnotateCtx(context.TODO())
+	ctx := ds.AnnotateCtx(stream.Context())
 	err := ds.flowStreamInt(stream)
 	if err != nil {
 		log.Error(ctx, err)

